### PR TITLE
PS Update for private link feature

### PR DESCRIPTION
**Pull request by bowen0106**
## Description
* Change all cmdlets for PrivateEndpointConnection to support generic service provider.
    - Updated cmdlet:
        - Approve-AzPrivateEndpointConnection
        - Deny-AzPrivateEndpointConnection
        - Get-AzPrivateEndpointConnection
        - Remove-AzPrivateEndpointConnection
        - Set-AzPrivateEndpointConnection
* Add new cmdlet for PrivateLinkResource and it also support generic service provider.
    - New cmdlet:
        - Get-AzPrivateLinkResource
* Add new fields and parameter for the feature Proxy Protocol V2.
    - Add property EnableProxyProtocol in PrivateLinkService
    - Add property LinkIdentifier in PrivateEndpointConnection
    - Updated New-AzPrivateLinkService to add a new optional parameter EnableProxyProtocol.

## Checklist

- [x] I have read the [_Submitting Changes_](../CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

File: src/Network/Network/PrivateLinkService/NewAzurePrivateLinkServiceCommand.cs [hunk 1]
```diff
@@ -117,6 +122,15 @@ private PSPrivateLinkService CreatePSPrivateLinkService()
                 psPrivateLinkService.AutoApproval.Subscriptions = AutoApproval.ToList();
             }
 
+            if (this.EnableProxyProtocol.IsPresent)
```
Comment (number213) [inline src/Network/Network/PrivateLinkService/NewAzurePrivateLinkServiceCommand.cs]: `psPrivateLinkService.EnableProxyProtocol = this.EnableProxyProtocol.IsPresent` would be cleaner

**Comment by wyunchi-ms** [comment]
Hi @bowen0106 . Since you are updating some cmdlets, please submit a design issue at [here](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues) first. If you already have one, please give me the link.
PS. Design complete date of Az 3.0.0 is 10.18. Please submit the design issue in today if you want this change in this release. We will review it as soon as possible.

File: src/Network/Network/Common/NetworkClient.cs [hunk 2]
```diff
@@ -45,6 +48,7 @@ public partial class NetworkClient
         public NetworkClient(IAzureContext context)
             : this(AzureSession.Instance.ClientFactory.CreateArmClient<NetworkManagementClient>(context, AzureEnvironment.Endpoint.ResourceManager))
         {
+            SqlManagementClient = AzureSession.Instance.ClientFactory.CreateArmClient<SqlManagementClient>(context, AzureEnvironment.Endpoint.ResourceManager);
```
Comment (number213) [inline src/Network/Network/Common/NetworkClient.cs]: This would be initialized in cmdlet that don't need it. Can you rewrite SqlManagerClient so it would be created in getter if it's null?

File: src/Network/Network/PrivateLinkService/PrivateEndpointConnection/ApproveAzurePrivateEndpointConnection.cs [hunk 3]
```diff
@@ -29,46 +29,49 @@ namespace Microsoft.Azure.Commands.Network
...
-
-            var psPrivateLinkService = this.GetPrivateLinkService(ResourceGroupName, ServiceName);
-            var peConnection = psPrivateLinkService.PrivateEndpointConnections.FirstOrDefault(
-                    resource =>
-                        string.Equals(resource.Name, this.Name, StringComparison.CurrentCultureIgnoreCase));
 
-            if (peConnection == null)
-            {
-                throw new ArgumentException(string.Format(Properties.Resources.ResourceNotFound, this.Name));
+                if (this.ResourceId.Split(new[] { '/' }, StringSplitOptions.RemoveEmptyEntries).Contains("privateEndpointConnections"))
```
Comment (number213) [inline src/Network/Network/PrivateLinkService/PrivateEndpointConnection/ApproveAzurePrivateEndpointConnection.cs]: You case check `resourceIdentifier.ResourceType` instead of splitting id string (applies to other similar code)

File: src/Network/Network/PrivateLinkService/PrivateEndpointConnection/GetAzurePrivateEndpointConnection.cs [hunk 4]
```diff
@@ -24,6 +24,7 @@
 using Microsoft.WindowsAzure.Commands.Utilities.Common;
 using Microsoft.Azure.Management.Internal.Resources.Utilities.Models;
 using System.Linq;
+using System.Threading.Tasks;
```
Comment (number213) [inline src/Network/Network/PrivateLinkService/PrivateEndpointConnection/GetAzurePrivateEndpointConnection.cs]: Do you need it here?

File: src/Network/Network/Common/NetworkClient.cs (hunk 2 above)
Comment (bowen0106) [reply src/Network/Network/Common/NetworkClient.cs]: Thanks for advice. Change for it for later initialization.

File: src/Network/Network/PrivateLinkService/PrivateEndpointConnection/ApproveAzurePrivateEndpointConnection.cs (hunk 3 above)
Comment (bowen0106) [reply src/Network/Network/PrivateLinkService/PrivateEndpointConnection/ApproveAzurePrivateEndpointConnection.cs]: Thanks for advice. I have realized the way for resourceIdentifier. The main controversy here is the parameter resourceId whether be resource id for private endpoint connection or for its parent resource (Sql/CosmosDB/PrivateLinkService). That's why previously I manually check the resource id to include the private endpoint connection there.
And from general rule for Powershell, the resource id should only be the resource that refer to the current cmdlet. So our PM and I reach an agreement for resource id still referring to private endpoint connection resource id.
So i removed the redundant code you are mentioned about. Thanks.

**Comment by wyunchi-ms** [comment]
Hi @bowen0106 , could you please take a look at the CI problem?

File: src/Network/Network/PrivateLinkService/PrivateLinkResource/GetAzurePrivateLinkResourceCommand.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+{
+    [Cmdlet("Get", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "PrivateLinkResource", DefaultParameterSetName = "ByResourceId"), OutputType(typeof(PSPrivateLinkResource))]
+    public class GetAzurePrivateLinkResourceCommand : NetworkBaseCmdlet
+    {
+        [Parameter(
+            Mandatory = true,
+            ParameterSetName = "ByResourceId",
+            ValueFromPipelineByPropertyName = true)]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
```
[the remark this answers is not in the document]
Comment (bowen0106) [reply src/Network/Network/PrivateLinkService/PrivateLinkResource/GetAzurePrivateLinkResourceCommand.cs]: @wyunchi-ms  Thanks for the advice. The thing is, `ResourceIdCompleter`  needs to take a resource type as parameter such as `[ResourceIdCompleter("Microsoft.Sql/servers")]`. However for here, we can accept **different type** of resources. So we cannot fix the resource type here.

**Comment by number213** [comment]
We've checked in test recordings with API version 2019-09-01 into `network-october`, please merge/rebase to unblock your PR

File: src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1 [hunk 6]
```diff
@@ -175,8 +175,11 @@ function Test-PrivateEndpointConnectionCRUD
         Assert-NotNull $pecApprove;
         Assert-AreEqual "Approved" $pecApprove.PrivateLinkServiceConnectionState.Status
 
+        Start-Sleep -s 30
```
Comment (number213) [inline src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1]: Please use `Start-TestSleep`. That way it won't sleep in playback mode

**Comment by wyunchi-ms** [comment]
Hi @bowen0106 could you please take a look at the CI?

File: src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1 [hunk 7]
```diff
@@ -118,3 +118,87 @@ function Test-PrivateLinkServiceCRUD
...
+                -ServerName $serverName `
+                -DatabaseName $databaseName `
+                -RequestedServiceObjectiveName "Basic" `
+                -Edition "Basic"
+
+            $sqlResourceId = $server.ResourceId
+        }
+        else
+        {
+            $sqlResourceId = "/subscriptions/e05dbbce-79c2-45a2-a7ef-f1058856feb3/resourceGroups/ps3301/providers/Microsoft.Sql/servers/server-841315352"
```
Comment (wyunchi-ms) [inline src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1]: I don't think it's a good idea to put usename, password, subscription and other sensitive info in the code.

Comment (bowen0106) [reply src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1]: @wyunchi-ms 
This is for testing purpose and it will create resource group and all other resource and delete after test.
Also, this same code  snippet is the same as in 
\azure-powershell\src\Sql\Sql.Test\ScenarioTests\Common.ps1
```
<#
.SYNOPSIS
Creates the basic test environment needed to perform the Sql data security tests - resource group, server and database
#>
function Create-BasicTestEnvironmentWithParams ($params, $location, $serverVersion)
{
	New-AzResourceGroup -Name $params.rgname -Location $location
	$serverName = $params.serverName
	$serverLogin = "testusername"
	<#[SuppressMessage("Microsoft.Security", "CS002:SecretInNextLine", Justification="Test passwords only valid for the duration of the test")]#>
	$serverPassword = "t357ingP@s5w0rd!Sec"
	$credentials = new-object System.Management.Automation.PSCredential($serverLogin, ($serverPassword | ConvertTo-SecureString -asPlainText -Force))
	New-AzSqlServer -ResourceGroupName $params.rgname -ServerName $params.serverName -Location $location -ServerVersion $serverVersion -SqlAdministratorCredentials $credentials
	New-AzSqlDatabase -DatabaseName $params.databaseName -ResourceGroupName $params.rgname -ServerName $params.serverName -Edition Basic
}
```
So I don't think there is any harm or concern on this passcode.
Thanks.

Comment (number213) [reply src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1]: This would work only once. Resource group name and server name are randomized, and the different subscription could be used. If you need sqlResourceId - construct it from variables

File: tools/StaticAnalysis/Exceptions/Az.Network/BreakingChangeIssues.csv [hunk 8]
```diff
@@ -502,4 +502,7 @@
...
\ No newline at end of file
+"Microsoft.Azure.PowerShell.Cmdlets.Network.dll","Microsoft.Azure.Commands.Network.NewAzureRmVpnSiteCommand","New-AzVpnSite","0","1050","The parameter set '__AllParameterSets' for cmdlet 'New-AzVpnSite' has been removed.","Add parameter set '__AllParameterSets' back to cmdlet 'New-AzVpnSite'."
+"Microsoft.Azure.PowerShell.Cmdlets.Network.dll","Microsoft.Azure.Commands.Network.SetAzurePrivateEndpointConnection","Set-AzPrivateEndpointConnection","0","1020","The cmdlet 'Set-AzPrivateEndpointConnection' no longer has output type 'Microsoft.Azure.Commands.Network.Models.PSPrivateLinkService'.","Make cmdlet 'Set-AzPrivateEndpointConnection' return type 'Microsoft.Azure.Commands.Network.Models.PSPrivateLinkService'."
+"Microsoft.Azure.PowerShell.Cmdlets.Network.dll","Microsoft.Azure.Commands.Network.GetAzurePrivateEndpointConnection","Get-AzPrivateEndpointConnection","0","2000","The cmdlet 'Get-AzPrivateEndpointConnection' no longer supports the parameter 'Description' and no alias was found for the original parameter name.","Add the parameter 'Description' back to the cmdlet 'Get-AzPrivateEndpointConnection', or add an alias to the original parameter name."
```
Comment (number213) [inline tools/StaticAnalysis/Exceptions/Az.Network/BreakingChangeIssues.csv]: Please remove these two entries since you revert the change to this property

File: src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1 [hunk 9]
```diff
@@ -24,6 +24,15 @@ function Check-CmdletReturnType
     return $cmdletReturnTypes -contains $realReturnType;
 }
 
+<#
+.SYNOPSIS
+Gets the subscription ID from the recorded mock for a test
+#>
+function getSubscription
```
Comment (number213) [inline src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1]: This function is already defined in `Common.ps1`. No need to add it again. Sorry for confusion.

File: src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1 (hunk 7 above)
Comment (wyunchi-ms) [reply src/Network/Network.Test/ScenarioTests/PrivateLinkServiceTests.ps1]: Got it.
